curl/tests/mock_curl.cc

```diff
@@ -0,0 +1,361 @@
+#define BUILDING_LIBCURL
+#include <cstring>
+#include <stdarg.h>
+#include "mock_curl.hpp"
+#include <iostream>
+
+using namespace std;
+
+/*
+ * libcurl implementations below. We are mocking necessary methods
+ * of the CURL API to ensure that our wrapper is making the calls
+ * to libcurl that we expect.
+ */
+
+/*
+ * Error struct implementations below... <for testing various libcurl failures>
+ * These are for the pre-client inits. The rest of the exceptions can use an
+ * enum inside of the curl_impl class
+ */
+error_mode test_failure_mode = success;
+
+curl_fail_init::curl_fail_init(error_mode mode)
+{
+   test_failure_mode = mode;
+}
+
+curl_fail_init::~curl_fail_init()
+{
+    test_failure_mode = success;
+}
+
+/*
+ * Sets up the program environment that libcurl needs. The mock
+ * implementation simply returns successfully.
+ */
+CURLcode curl_global_init(long flags)
+{
+    if (test_failure_mode == global_init_error) {
+        return CURLE_FAILED_INIT;
+    } else {
+        return CURLE_OK;
+    }
+}
+
+/*
+ * Reclaim memory obtained from a libcurl call by deleting
+ * a mock curl object.
+ */
+void curl_free(void *p)
+{
+    delete static_cast<curl_impl*>(p);
+}
+
+/*
+ * End a libcurl easy handle. The mock implementation simply
+ * calls curl_free to delete the mock curl object argument.
+ */
+void curl_easy_cleanup(CURL * handle)
+{
+    curl_free(handle);
+}
+
+/*
+ * Mock implementation of curl_easy_escape which simply returns a
+ * nullptr. URL encoding the given string is not necessary for
+ * testing.
+ */
+char *curl_easy_escape(CURL * curl, const char * string, int length)
+{
+    return nullptr;
+}
+
+/*
+ * Start a libcurl easy session. The mock implementation simply returns
+ * a new mock curl object.
+ */
+CURL *curl_easy_init()
+{
+    if (test_failure_mode == easy_init_error) {
+        return nullptr;
+    } else {
+        return new curl_impl();
+    }
+}
+
+/*
+ * Set options for an easy curl handle. We use this method in the mock
+ * implementation to ensure that the correct CURL API calls are being
+ * made. Given a particular CURL option, we store received data from
+ * the varargs parameter in the mock curl object. The data is
+ * then verified in the tests.
+ */
+CURLcode curl_easy_setopt(CURL *handle, CURLoption option, ...)
+{
+    auto h = static_cast<curl_impl*>(handle);
+    va_list vl;
+    va_start(vl, option);
+
+    switch (option) {
+        case CURLOPT_HEADERFUNCTION:
+            // Set client::write_header as the function to be called as soon as mock curl has received header data.
+            if (h->test_failure_mode == curl_impl::error_mode::header_function_error) {
```

Comment:
I don't see anything in the [curl documentation](http://curl.haxx.se/libcurl/c/CURLOPT_HEADERFUNCTION.html) saying this will return an error.